zend_abstract_interface/symbols/lookup.c

```diff
@@ -0,0 +1,519 @@
+#include <sandbox/sandbox.h>
+#include "symbols.h"
+#include "zend_compile.h"
+
+// clang-format off
+static inline bool zai_symbol_update(zend_class_entry *ce ZAI_TSRMLS_DC) {
+    if (ce->ce_flags & ZEND_ACC_CONSTANTS_UPDATED) {
+        return true;
+    }
+
+    volatile bool zai_symbol_updated = true;
+
+    zai_sandbox sandbox;
+    zai_sandbox_open(&sandbox);
+
+    zend_try {
+#if PHP_VERSION_ID < 70000
+        zend_update_class_constants(ce ZAI_TSRMLS_CC);
+#else
+        if (zend_update_class_constants(ce) != SUCCESS) {
+            zai_symbol_updated = false;
+        }
+#endif
+    } zend_catch {
+        zai_symbol_updated = false;
+    } zend_end_try();
+
+    if (zai_symbol_updated && !EG(exception)) {
+        zai_sandbox_close(&sandbox);
+        return true;
+    }
+
+    zai_sandbox_close(&sandbox);
+    return false;
+}
+
+static inline void *zai_symbol_lookup_table(HashTable *table, const char *name, size_t length, bool ncase, bool pointer ZAI_TSRMLS_DC) {
+#if PHP_VERSION_ID >= 70000
+    zval *result = zend_hash_str_find(table, name, length);
+#else
+    void *result = NULL;
+
+    zend_hash_find(table, name, length + 1, (void **)&result);
+#endif
+
+    if (!result && ncase) {
+        char *ptr = (char *)pemalloc(length + 1, 1);
+
+        for (uint32_t c = 0; c < length; c++) {
+            ptr[c] = tolower(name[c]);
+        }
+
+        ptr[length] = 0;
+
+#if PHP_VERSION_ID >= 70000
+        result = zend_hash_str_find(table, ptr, length);
+#else
+        zend_hash_find(table, ptr, length + 1, (void **)&result);
+#endif
+
+        pefree(ptr, 1);
+    }
+
+#if PHP_VERSION_ID >= 70000
+    if (pointer && result) {
+        return Z_PTR_P(result);
+    } else {
+        return result;
+    }
+#endif
+    return result;
+}
+
+static inline size_t zai_symbol_lookup_lengthof(zai_string_view *view) {
+    if (!view->len) {
+        return 0;
+    }
+
+    if (memcmp(view->ptr, "\\", sizeof("\\")-1) == SUCCESS) {
+        return view->len - 1;
+    }
+
+    return view->len;
+}
+
+static inline const char* zai_symbol_lookup_startof(zai_string_view *view) {
+    if (!view->len) {
+        return NULL;
+    }
+
+    if (memcmp(view->ptr, "\\", sizeof("\\")-1) == SUCCESS) {
+        return view->ptr + 1;
+    }
+
+    return view->ptr;
+}
+
+static inline zai_string_view* zai_symbol_lookup_key(zai_string_view *namespace, zai_string_view *name, bool lower ZAI_TSRMLS_DC) {
+    zai_string_view *result = pemalloc(sizeof(zai_string_view), 1);
```

Comment:
```suggestion
static inline zai_string_view zai_symbol_lookup_key(zai_string_view *namespace, zai_string_view *name, bool lower ZAI_TSRMLS_DC) {
    zai_string_view result = pemalloc(sizeof(zai_string_view), 1);
```
don't alloc zai_string_views, return it without pointers. It's a 16 byte structure. A compiler can efficiently transfer that, an alloc is costly.

Reply:
done